File: python/tests/test_host_calls.py
"""Tests for the child→host call surface (memory.recall / llm.complete).

Each subprocess test spawns a tiny driver that imports
``nexo_plugin_sdk.PluginAdapter`` and, in its ``on_event`` handler,
makes a host call (``broker.memory_recall`` / ``broker.llm_complete`` /
``broker.llm_complete_stream``). The test then acts as the *host*: it
feeds the child a ``broker.event``, reads the child's request frame off
the child's stdout, writes a canned response to the child's stdin, and
asserts what the handler did (it publishes a marker derived from the
host's reply). Mirrors the spawn-driver style of the other test files.
"""

import json
import os
import select
import subprocess
import sys
import time
import unittest
from pathlib import Path

from nexo_plugin_sdk.host import LlmCompleteResult, MemoryEntry

SDK_ROOT = Path(__file__).resolve().parent.parent

# One driver covers the single-call cases; the event payload's `op`
# field selects which host call the handler makes.
DRIVER_HOST_CALL = r"""
import asyncio, sys
from nexo_plugin_sdk import PluginAdapter, Event
from nexo_plugin_sdk.errors import RpcServerError, RpcTimeoutError, RpcTransportError

MANIFEST = '''
[plugin]
id = "host_call_plugin"
version = "0.1.0"
name = "HostCall"
description = "fixture"
min_nexo_version = ">=0.1.0"
'''

async def on_event(topic, event, broker):
    op = event.payload.get("op")
    try:
        if op == "recall":
            entries = await broker.memory_recall(agent_id="agent_x", query="prefs", limit=3)
            await broker.publish("plugin.inbound.out", Event.new("plugin.inbound.out", "host_call_plugin",
                {"got": [e.content for e in entries], "ids": [e.id for e in entries]}))
        elif op == "complete":
            r = await broker.llm_complete(provider="minimax", model="m", messages=[{"role": "user", "content": "hi"}])
            await broker.publish("plugin.inbound.out", Event.new("plugin.inbound.out", "host_call_plugin",
                {"content": r.content, "finish_reason": r.finish_reason, "ct": r.usage.completion_tokens}))
        elif op == "stream":
            stream = broker.llm_complete_stream(provider="minimax", model="m", messages=[{"role": "user", "content": "hi"}])
            chunks = []
            async for c in stream:
                chunks.append(c)
            final = await stream.final_result()
            await broker.publish("plugin.inbound.out", Event.new("plugin.inbound.out", "host_call_plugin",
                {"chunks": chunks, "joined": "".join(chunks), "content_is_none": final.content is None,
                 "finish_reason": final.finish_reason}))
        elif op == "timeout":
            try:
                await broker.memory_recall(agent_id="a", query="q", timeout=0.4)
            except RpcTimeoutError as e:
                await broker.publish("plugin.inbound.out", Event.new("plugin.inbound.out", "host_call_plugin",
                    {"timed_out": True, "seconds": e.seconds}))
        elif op == "error":
            try:
                await broker.memory_recall(agent_id="a", query="q")
            except RpcServerError as e:
                await broker.publish("plugin.inbound.out", Event.new("plugin.inbound.out", "host_call_plugin",
                    {"rpc_error_code": e.code, "rpc_error_msg": e.message}))
        elif op == "inflight":
            # Never replied to — the adapter abandons it on shutdown.
            try:
                await broker.memory_recall(agent_id="a", query="q")
            except RpcTransportError as e:
                sys.stderr.write(f"HANDLER_GOT_TRANSPORT: {e}\n"); sys.stderr.flush()
    except Exception as e:
        sys.stderr.write(f"HANDLER_RAISED {type(e).__name__}: {e}\n"); sys.stderr.flush()

async def main():
    await PluginAdapter(manifest_toml=MANIFEST, on_event=on_event).run()

asyncio.run(main())
"""

# Two-event driver for the out-of-order multiplexing test: each event
# issues its own memory.recall with a distinct query echoed back.
DRIVER_MULTIPLEX = r"""
import asyncio
from nexo_plugin_sdk import PluginAdapter, Event

MANIFEST = '''
[plugin]
id = "mux_plugin"
version = "0.1.0"
name = "Mux"
description = "fixture"
min_nexo_version = ">=0.1.0"
'''

async def on_event(topic, event, broker):
    q = event.payload["q"]
    entries = await broker.memory_recall(agent_id="a", query=q)
    await broker.publish("plugin.inbound.out", Event.new("plugin.inbound.out", "mux_plugin",
        {"q": q, "got": [e.content for e in entries]}))

async def main():
    await PluginAdapter(manifest_toml=MANIFEST, on_event=on_event).run()

asyncio.run(main())
"""


def _kill_if_alive(proc) -> None:
    if proc.poll() is None:
        proc.kill()
        proc.wait()


def _spawn(self, driver: str) -> "subprocess.Popen[bytes]":
    env = dict(os.environ)
    env["PYTHONPATH"] = str(SDK_ROOT) + os.pathsep + env.get("PYTHONPATH", "")
    p = subprocess.Popen(
        [sys.executable, "-c", driver],
        stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=env,
    )
    p._framebuf = b""  # type: ignore[attr-defined]
    self.addCleanup(_kill_if_alive, p)
    return p


def _write_frame(proc, frame: dict) -> None:
    proc.stdin.write((json.dumps(frame) + "\n").encode("utf-8"))
    proc.stdin.flush()


def _read_frame(proc, timeout: float = 5.0) -> dict:
    """Read the next JSON-RPC line off the child's stdout (buffered, so a
    single OS read that grabs several frames is handled), or fail."""
    deadline = time.monotonic() + timeout
    while True:
        if b"\n" in proc._framebuf:
            line, _, proc._framebuf = proc._framebuf.partition(b"\n")
            return json.loads(line.decode("utf-8"))
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            raise AssertionError(f"no frame from child within {timeout}s (buf={proc._framebuf!r})")
        r, _, _ = select.select([proc.stdout], [], [], min(0.1, remaining))
        if r:
            chunk = os.read(proc.stdout.fileno(), 65536)
            if not chunk:
                raise AssertionError(f"child stdout closed (buf={proc._framebuf!r})")
            proc._framebuf += chunk


def _expect_exit_0(proc) -> None:
    rc = proc.wait(timeout=5)
    if rc != 0:
        raise AssertionError(f"child exited {rc}; stderr={proc.stderr.read().decode('utf-8', 'replace')!r}")


def _event_frame(topic: str, payload: dict) -> dict:
    return {"jsonrpc": "2.0", "method": "broker.event",
            "params": {"topic": topic, "event": {"topic": topic, "source": "host", "payload": payload}}}


def _shutdown_frame(req_id: int = 99) -> dict:
    return {"jsonrpc": "2.0", "id": req_id, "method": "shutdown"}


class HostTypeTests(unittest.TestCase):
    def test_memory_entry_from_json(self):
        e = MemoryEntry.from_json({"id": "x1", "agent_id": "a", "content": "c", "tags": ["t"],
                                   "concept_tags": [], "created_at": "2026-01-01T00:00:00Z", "memory_type": None})
        self.assertEqual((e.id, e.agent_id, e.content, e.tags, e.memory_type), ("x1", "a", "c", ["t"], None))

    def test_llm_result_from_json(self):
        r = LlmCompleteResult.from_json({"content": "hi", "finish_reason": "stop",
                                         "usage": {"prompt_tokens": 5, "completion_tokens": 2}})
        self.assertEqual((r.content, r.finish_reason, r.usage.prompt_tokens, r.usage.completion_tokens), ("hi", "stop", 5, 2))
        r2 = LlmCompleteResult.from_json({"finish_reason": "length"})  # streamed final reply — no content
        self.assertIsNone(r2.content)
        self.assertEqual(r2.finish_reason, "length")


class MemoryRecallTests(unittest.TestCase):
    def test_recall_happy_path(self):
        proc = _spawn(self, DRIVER_HOST_CALL)
        _write_frame(proc, _event_frame("plugin.in", {"op": "recall"}))
        req = _read_frame(proc)
        self.assertEqual(req["method"], "memory.recall")
        self.assertEqual(req["params"], {"agent_id": "agent_x", "query": "prefs", "limit": 3})
        _write_frame(proc, {"jsonrpc": "2.0", "id": req["id"], "result": {"entries": [
            {"id": "m1", "agent_id": "agent_x", "content": "concise", "tags": ["pref"], "concept_tags": [],
             "created_at": "2026-01-01T00:00:00Z", "memory_type": None}]}})
        out = _read_frame(proc)
        self.assertEqual(out["method"], "broker.publish")
        self.assertEqual(out["params"]["event"]["payload"], {"got": ["concise"], "ids": ["m1"]})
        _write_frame(proc, _shutdown_frame())
        self.assertTrue(_read_frame(proc)["result"]["ok"])
        _expect_exit_0(proc)

    def test_server_error_surfaces_as_rpc_server_error(self):
        proc = _spawn(self, DRIVER_HOST_CALL)
        _write_frame(proc, _event_frame("plugin.in", {"op": "error"}))
        req = _read_frame(proc)
        self.assertEqual(req["method"], "memory.recall")
        _write_frame(proc, {"jsonrpc": "2.0", "id": req["id"], "error": {"code": -32603, "message": "memory not configured"}})
        out = _read_frame(proc)
        self.assertEqual(out["params"]["event"]["payload"], {"rpc_error_code": -32603, "rpc_error_msg": "memory not configured"})
        _write_frame(proc, _shutdown_frame())
        self.assertTrue(_read_frame(proc)["result"]["ok"])
        _expect_exit_0(proc)

    def test_timeout_surfaces_as_rpc_timeout_error(self):
        proc = _spawn(self, DRIVER_HOST_CALL)
        _write_frame(proc, _event_frame("plugin.in", {"op": "timeout"}))
        req = _read_frame(proc)
        self.assertEqual(req["method"], "memory.recall")
        # Deliberately never reply — the 0.4s per-call timeout fires.
        out = _read_frame(proc, timeout=5.0)
        self.assertTrue(out["params"]["event"]["payload"]["timed_out"])
        self.assertAlmostEqual(out["params"]["event"]["payload"]["seconds"], 0.4, places=2)
        # Dispatch still alive — a late reply for the timed-out id is dropped; shutdown still works.
        _write_frame(proc, {"jsonrpc": "2.0", "id": req["id"], "result": {"entries": []}})
        _write_frame(proc, _shutdown_frame())
        self.assertTrue(_read_frame(proc)["result"]["ok"])
        _expect_exit_0(proc)

    def test_late_response_for_unknown_id_dropped(self):
        proc = _spawn(self, DRIVER_HOST_CALL)
        _write_frame(proc, {"jsonrpc": "2.0", "id": 9999, "result": {"entries": []}})  # never issued — must not crash
        _write_frame(proc, _shutdown_frame())
        self.assertTrue(_read_frame(proc)["result"]["ok"])
        rc = proc.wait(timeout=5)
        self.assertEqual(rc, 0)
        self.assertIn("unknown/expired request id", proc.stderr.read().decode("utf-8"))


class LlmCompleteTests(unittest.TestCase):
    def test_complete_non_stream_happy_path(self):
        proc = _spawn(self, DRIVER_HOST_CALL)
        _write_frame(proc, _event_frame("plugin.in", {"op": "complete"}))
        req = _read_frame(proc)
        self.assertEqual(req["method"], "llm.complete")
        self.assertEqual(req["params"]["provider"], "minimax")
        self.assertNotIn("stream", req["params"])
        _write_frame(proc, {"jsonrpc": "2.0", "id": req["id"], "result": {
            "content": "one line.", "finish_reason": "stop", "usage": {"prompt_tokens": 3, "completion_tokens": 4}}})
        out = _read_frame(proc)
        self.assertEqual(out["params"]["event"]["payload"], {"content": "one line.", "finish_reason": "stop", "ct": 4})
        _write_frame(proc, _shutdown_frame())
        self.assertTrue(_read_frame(proc)["result"]["ok"])
        _expect_exit_0(proc)

    def test_complete_streaming_happy_path(self):
        proc = _spawn(self, DRIVER_HOST_CALL)
        _write_frame(proc, _event_frame("plugin.in", {"op": "stream"}))
        req = _read_frame(proc)
        self.assertEqual(req["method"], "llm.complete")
        self.assertTrue(req["params"]["stream"])
        rid = req["id"]
        for c in ("hel", "lo ", "world"):
            _write_frame(proc, {"jsonrpc": "2.0", "method": "llm.complete.delta", "params": {"request_id": rid, "chunk": c}})
        _write_frame(proc, {"jsonrpc": "2.0", "id": rid, "result": {"finish_reason": "stop", "usage": {"prompt_tokens": 1, "completion_tokens": 3}}})
        out = _read_frame(proc)
        p = out["params"]["event"]["payload"]
        self.assertEqual(p["chunks"], ["hel", "lo ", "world"])
        self.assertEqual(p["joined"], "hello world")
        self.assertTrue(p["content_is_none"])
        self.assertEqual(p["finish_reason"], "stop")
        _write_frame(proc, _shutdown_frame())
        self.assertTrue(_read_frame(proc)["result"]["ok"])
        _expect_exit_0(proc)


class MultiplexingTests(unittest.TestCase):
    def test_two_in_flight_calls_resolve_out_of_order(self):
        proc = _spawn(self, DRIVER_MULTIPLEX)
        _write_frame(proc, _event_frame("plugin.in", {"q": "alpha"}))
        _write_frame(proc, _event_frame("plugin.in", {"q": "beta"}))
        req1 = _read_frame(proc)
        req2 = _read_frame(proc)
        qs = {req1["params"]["query"]: req1["id"], req2["params"]["query"]: req2["id"]}
        self.assertEqual(set(qs), {"alpha", "beta"})
        # Reply to BETA first, then ALPHA.
        _write_frame(proc, {"jsonrpc": "2.0", "id": qs["beta"], "result": {"entries": [{"id": "b", "agent_id": "a", "content": "B-content"}]}})
        _write_frame(proc, {"jsonrpc": "2.0", "id": qs["alpha"], "result": {"entries": [{"id": "a", "agent_id": "a", "content": "A-content"}]}})
        out_a = _read_frame(proc)
        out_b = _read_frame(proc)
        by_q = {o["params"]["event"]["payload"]["q"]: o["params"]["event"]["payload"]["got"] for o in (out_a, out_b)}
        self.assertEqual(by_q, {"alpha": ["A-content"], "beta": ["B-content"]})
        _write_frame(proc, _shutdown_frame())
        self.assertTrue(_read_frame(proc)["result"]["ok"])
        _expect_exit_0(proc)


class ShutdownWhileInFlightTests(unittest.TestCase):
    def test_shutdown_while_host_call_in_flight_does_not_hang(self):
        proc = _spawn(self, DRIVER_HOST_CALL)
        _write_frame(proc, _event_frame("plugin.in", {"op": "inflight"}))
        req = _read_frame(proc)
        self.assertEqual(req["method"], "memory.recall")
        # Don't reply — shut down. The adapter must abandon the in-flight
        # call (handler gets RpcTransportError) and reply {ok:true}
        # promptly, not hang waiting for a response that won't come.
        _write_frame(proc, _shutdown_frame())
        self.assertTrue(_read_frame(proc, timeout=5.0)["result"]["ok"])
        rc = proc.wait(timeout=5)
        self.assertEqual(rc, 0)
        self.assertIn("HANDLER_GOT_TRANSPORT", proc.stderr.read().decode("utf-8"))


if __name__ == "__main__":
    unittest.main()

File: php/tests/test_host_calls.php
<?php

declare(strict_types=1);

/**
 * Tests for the child→host call surface (memory.recall / llm.complete).
 *
 * The fixture makes the host call in onEvent; this test plays the *host*
 * — it spawns the fixture, feeds it a broker.event, reads the child's
 * request frame off stdout, writes a canned reply to stdin, and asserts
 * the published-back marker. Interactive (interleaved reads/writes), so
 * it uses its own non-blocking harness rather than run_fixture().
 */

require __DIR__ . '/../vendor/autoload.php';

const FIXTURE = __DIR__ . '/fixtures/host-call-plugin.php';

function fail(string $msg): never
{
    fwrite(STDERR, "FAIL: $msg\n");
    exit(1);
}

/** @return array{proc: resource, in: resource, out: resource, err: resource, buf: string} */
function spawn_fixture(): array
{
    $proc = proc_open(
        ['php', FIXTURE],
        [0 => ['pipe', 'r'], 1 => ['pipe', 'w'], 2 => ['pipe', 'w']],
        $pipes,
    );
    if (!is_resource($proc)) {
        fail('proc_open failed');
    }
    stream_set_blocking($pipes[1], false);
    stream_set_blocking($pipes[2], false);
    return ['proc' => $proc, 'in' => $pipes[0], 'out' => $pipes[1], 'err' => $pipes[2], 'buf' => ''];
}

function write_frame(array &$ctx, array $frame): void
{
    fwrite($ctx['in'], json_encode($frame) . "\n");
    fflush($ctx['in']);
}

/** Read the next JSON-RPC line off the child's stdout (buffered). */
function read_frame(array &$ctx, float $timeoutSec = 5.0): array
{
    $deadline = microtime(true) + $timeoutSec;
    while (true) {
        $nl = strpos($ctx['buf'], "\n");
        if ($nl !== false) {
            $line = substr($ctx['buf'], 0, $nl);
            $ctx['buf'] = substr($ctx['buf'], $nl + 1);
            if (trim($line) === '') {
                continue;
            }
            $f = json_decode($line, true);
            if (!is_array($f)) {
                fail('read_frame: non-object line: ' . $line);
            }
            return $f;
        }
        if (microtime(true) >= $deadline) {
            fail('read_frame: no frame within ' . $timeoutSec . 's (buf=' . $ctx['buf'] . ')');
        }
        $chunk = stream_get_contents($ctx['out']);
        if ($chunk !== false && $chunk !== '') {
            $ctx['buf'] .= $chunk;
        } else {
            usleep(5_000);
        }
    }
}

function expect_exit_0(array &$ctx, float $timeoutSec = 5.0): string
{
    $deadline = microtime(true) + $timeoutSec;
    $exitCode = null;
    while (microtime(true) < $deadline) {
        $status = proc_get_status($ctx['proc']);
        if (!$status['running']) {
            $exitCode = $status['exitcode']; // proc_get_status reaped it; proc_close would return -1
            break;
        }
        usleep(5_000);
    }
    $err = stream_get_contents($ctx['err']) ?: '';
    @fclose($ctx['in']);
    @fclose($ctx['out']);
    @fclose($ctx['err']);
    proc_close($ctx['proc']);
    if ($exitCode !== 0) {
        fail('child exited ' . var_export($exitCode, true) . '; stderr=' . $err);
    }
    return $err;
}

function kill_proc(array &$ctx): void
{
    $status = @proc_get_status($ctx['proc']);
    if (is_array($status) && $status['running']) {
        @proc_terminate($ctx['proc'], 9);
    }
    @fclose($ctx['in']);
    @fclose($ctx['out']);
    @fclose($ctx['err']);
    @proc_close($ctx['proc']);
}

function read_stderr(array &$ctx): string
{
    return stream_get_contents($ctx['err']) ?: '';
}

function event_frame(string $topic, array $payload): array
{
    return ['jsonrpc' => '2.0', 'method' => 'broker.event',
        'params' => ['topic' => $topic, 'event' => ['topic' => $topic, 'source' => 'host', 'payload' => $payload]]];
}
function shutdown_frame(int $id = 99): array
{
    return ['jsonrpc' => '2.0', 'id' => $id, 'method' => 'shutdown'];
}

$n = 0;
$ok = function (string $name) use (&$n): void {
    $n++;
    fwrite(STDOUT, "ok $n - $name\n");
};

// ── data type fromJson round-trips ─────────────────────────────
$e = \Nexo\Plugin\Sdk\MemoryEntry::fromJson(['id' => 'x1', 'agent_id' => 'a', 'content' => 'c', 'tags' => ['t'], 'concept_tags' => [], 'created_at' => '2026-01-01T00:00:00Z', 'memory_type' => null]);
if ([$e->id, $e->agentId, $e->content, $e->tags, $e->memoryType] !== ['x1', 'a', 'c', ['t'], null]) {
    fail('MemoryEntry::fromJson: ' . print_r($e, true));
}
$r = \Nexo\Plugin\Sdk\LlmCompleteResult::fromJson(['content' => 'hi', 'finish_reason' => 'stop', 'usage' => ['prompt_tokens' => 5, 'completion_tokens' => 2]]);
if ([$r->content, $r->finishReason, $r->usage->promptTokens, $r->usage->completionTokens] !== ['hi', 'stop', 5, 2]) {
    fail('LlmCompleteResult::fromJson: ' . print_r($r, true));
}
$r2 = \Nexo\Plugin\Sdk\LlmCompleteResult::fromJson(['finish_reason' => 'length']);
if ($r2->content !== null || $r2->finishReason !== 'length') {
    fail('LlmCompleteResult::fromJson streamed: ' . print_r($r2, true));
}
$ok('data_type_fromJson_roundtrips');

// ── memory.recall happy path ───────────────────────────────────
$ctx = spawn_fixture();
write_frame($ctx, event_frame('plugin.in', ['op' => 'recall']));
$req = read_frame($ctx);
if (($req['method'] ?? null) !== 'memory.recall') {
    kill_proc($ctx);
    fail('memory_recall_happy: expected memory.recall, got ' . json_encode($req));
}
if ($req['params'] !== ['agent_id' => 'agent_x', 'query' => 'prefs', 'limit' => 3]) {
    kill_proc($ctx);
    fail('memory_recall_happy: wrong params ' . json_encode($req['params']));
}
write_frame($ctx, ['jsonrpc' => '2.0', 'id' => $req['id'], 'result' => ['entries' => [
    ['id' => 'm1', 'agent_id' => 'agent_x', 'content' => 'concise', 'tags' => ['pref'], 'concept_tags' => [], 'created_at' => '2026-01-01T00:00:00Z', 'memory_type' => null]]]]);
$out = read_frame($ctx);
if (($out['method'] ?? null) !== 'broker.publish' || ($out['params']['event']['payload'] ?? null) !== ['got' => ['concise'], 'ids' => ['m1']]) {
    kill_proc($ctx);
    fail('memory_recall_happy: wrong publish ' . json_encode($out));
}
write_frame($ctx, shutdown_frame());
if ((read_frame($ctx)['result']['ok'] ?? null) !== true) {
    kill_proc($ctx);
    fail('memory_recall_happy: shutdown reply not ok');
}
expect_exit_0($ctx);
$ok('memory_recall_happy_path');

// ── host error → RpcServerError ────────────────────────────────
$ctx = spawn_fixture();
write_frame($ctx, event_frame('plugin.in', ['op' => 'error']));
$req = read_frame($ctx);
write_frame($ctx, ['jsonrpc' => '2.0', 'id' => $req['id'], 'error' => ['code' => -32603, 'message' => 'memory not configured']]);
$out = read_frame($ctx);
if (($out['params']['event']['payload'] ?? null) !== ['rpc_error_code' => -32603, 'rpc_error_msg' => 'memory not configured']) {
    kill_proc($ctx);
    fail('server_error: wrong publish ' . json_encode($out));
}
write_frame($ctx, shutdown_frame());
read_frame($ctx);
expect_exit_0($ctx);
$ok('host_error_surfaces_as_RpcServerError');

// ── per-call timeout → RpcTimeoutError; dispatch survives ──────
$ctx = spawn_fixture();
write_frame($ctx, event_frame('plugin.in', ['op' => 'timeout']));
$req = read_frame($ctx);
if (($req['method'] ?? null) !== 'memory.recall') {
    kill_proc($ctx);
    fail('timeout: expected memory.recall');
}
// Never reply — the 0.4s per-call timeout fires.
$out = read_frame($ctx, 5.0);
if (($out['params']['event']['payload']['timed_out'] ?? null) !== true) {
    kill_proc($ctx);
    fail('timeout: expected timed_out marker, got ' . json_encode($out));
}
// A late reply for the timed-out id is dropped; shutdown still works.
write_frame($ctx, ['jsonrpc' => '2.0', 'id' => $req['id'], 'result' => ['entries' => []]]);
write_frame($ctx, shutdown_frame());
read_frame($ctx);
expect_exit_0($ctx);
$ok('per_call_timeout_surfaces_as_RpcTimeoutError');

// ── late response for unknown id dropped ───────────────────────
$ctx = spawn_fixture();
write_frame($ctx, ['jsonrpc' => '2.0', 'id' => 9999, 'result' => ['entries' => []]]); // never issued — must not crash
write_frame($ctx, shutdown_frame());
if ((read_frame($ctx)['result']['ok'] ?? null) !== true) {
    kill_proc($ctx);
    fail('late_response: shutdown reply not ok');
}
expect_exit_0($ctx);
$ok('late_response_for_unknown_id_dropped');

// ── llm.complete non-streaming happy path ──────────────────────
$ctx = spawn_fixture();
write_frame($ctx, event_frame('plugin.in', ['op' => 'complete']));
$req = read_frame($ctx);
if (($req['method'] ?? null) !== 'llm.complete' || ($req['params']['provider'] ?? null) !== 'minimax' || isset($req['params']['stream'])) {
    kill_proc($ctx);
    fail('llm_complete: wrong request ' . json_encode($req));
}
write_frame($ctx, ['jsonrpc' => '2.0', 'id' => $req['id'], 'result' => ['content' => 'one line.', 'finish_reason' => 'stop', 'usage' => ['prompt_tokens' => 3, 'completion_tokens' => 4]]]);
$out = read_frame($ctx);
if (($out['params']['event']['payload'] ?? null) !== ['content' => 'one line.', 'finish_reason' => 'stop', 'ct' => 4]) {
    kill_proc($ctx);
    fail('llm_complete: wrong publish ' . json_encode($out));
}
write_frame($ctx, shutdown_frame());
read_frame($ctx);
expect_exit_0($ctx);
$ok('llm_complete_non_streaming_happy_path');

// ── llm.complete streaming happy path ──────────────────────────
$ctx = spawn_fixture();
write_frame($ctx, event_frame('plugin.in', ['op' => 'stream']));
$req = read_frame($ctx);
if (($req['method'] ?? null) !== 'llm.complete' || ($req['params']['stream'] ?? null) !== true) {
    kill_proc($ctx);
    fail('llm_stream: wrong request ' . json_encode($req));
}
$rid = $req['id'];
foreach (['hel', 'lo ', 'world'] as $c) {
    write_frame($ctx, ['jsonrpc' => '2.0', 'method' => 'llm.complete.delta', 'params' => ['request_id' => $rid, 'chunk' => $c]]);
}
write_frame($ctx, ['jsonrpc' => '2.0', 'id' => $rid, 'result' => ['finish_reason' => 'stop', 'usage' => ['prompt_tokens' => 1, 'completion_tokens' => 3]]]);
$out = read_frame($ctx);
$p = $out['params']['event']['payload'] ?? [];
if (($p['chunks'] ?? null) !== ['hel', 'lo ', 'world'] || ($p['joined'] ?? null) !== 'hello world' || ($p['content_is_null'] ?? null) !== true || ($p['finish_reason'] ?? null) !== 'stop') {
    kill_proc($ctx);
    fail('llm_stream: wrong publish ' . json_encode($out));
}
write_frame($ctx, shutdown_frame());
read_frame($ctx);
expect_exit_0($ctx);
$ok('llm_complete_streaming_happy_path');

// ── two in-flight memory.recall calls resolve out of order ─────
$ctx = spawn_fixture();
write_frame($ctx, event_frame('plugin.in', ['q' => 'alpha']));
write_frame($ctx, event_frame('plugin.in', ['q' => 'beta']));
$req1 = read_frame($ctx);
$req2 = read_frame($ctx);
$ids = [$req1['params']['query'] => $req1['id'], $req2['params']['query'] => $req2['id']];
ksort($ids);
if (array_keys($ids) !== ['alpha', 'beta']) {
    kill_proc($ctx);
    fail('multiplex: unexpected queries ' . json_encode($ids));
}
// Reply to BETA first, then ALPHA.
write_frame($ctx, ['jsonrpc' => '2.0', 'id' => $ids['beta'], 'result' => ['entries' => [['id' => 'b', 'agent_id' => 'a', 'content' => 'B-content']]]]);
write_frame($ctx, ['jsonrpc' => '2.0', 'id' => $ids['alpha'], 'result' => ['entries' => [['id' => 'a', 'agent_id' => 'a', 'content' => 'A-content']]]]);
$o1 = read_frame($ctx);
$o2 = read_frame($ctx);
$byQ = [];
foreach ([$o1, $o2] as $o) {
    $byQ[$o['params']['event']['payload']['q']] = $o['params']['event']['payload']['got'];
}
ksort($byQ);
if ($byQ !== ['alpha' => ['A-content'], 'beta' => ['B-content']]) {
    kill_proc($ctx);
    fail('multiplex: cross-talk ' . json_encode($byQ));
}
write_frame($ctx, shutdown_frame());
read_frame($ctx);
expect_exit_0($ctx);
$ok('two_in_flight_calls_resolve_out_of_order');

// ── shutdown while a host call is in flight does not hang ──────
$ctx = spawn_fixture();
write_frame($ctx, event_frame('plugin.in', ['op' => 'inflight']));
$req = read_frame($ctx);
if (($req['method'] ?? null) !== 'memory.recall') {
    kill_proc($ctx);
    fail('inflight: expected memory.recall');
}
// Don't reply — shut down. The adapter must abandon the in-flight call
// (handler gets RpcTransportError) and reply {ok:true} promptly.
write_frame($ctx, shutdown_frame());
if ((read_frame($ctx, 5.0)['result']['ok'] ?? null) !== true) {
    kill_proc($ctx);
    fail('inflight: shutdown reply not ok / hung');
}
$err = expect_exit_0($ctx);
if (!str_contains($err, 'HANDLER_GOT_TRANSPORT')) {
    fail('inflight: handler should have caught RpcTransportError; stderr=' . $err);
}
$ok('shutdown_while_host_call_in_flight_does_not_hang');

exit(0);
